### ISSUE TITLE
ConfigurationGatheringTest failing locally but not on CI (and it should!)

### ISSUE THREAD

**Issue by fneu** [coala-analyzer/coala#2210]
So, ConfigurationGatheringTest fails on current master for me and @abhsag24 and possibly others.

Looking into it I found the following information:

# TL;DR:

```
>>> sections, local_bears, global_bears, targets = gather_configuration(lambda *args: True, log_printer, arg_list=[])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/fabian/devel/coala/coalib/settings/ConfigurationGathering.py", line 311, in gather_configuration
    log_printer)
  File "/home/fabian/devel/coala/coalib/settings/SectionFilling.py", line 38, in fill_settings
    fill_section(section, acquire_settings, log_printer, all_bears)
  File "/home/fabian/devel/coala/coalib/settings/SectionFilling.py", line 91, in fill_section
    for setting, help_text in new_vals.items():
AttributeError: 'bool' object has no attribute 'items'
```
# Analysis

`test_default_coafile_parsing(self):` in `ConfigurationGatheringTest.py` for example contains:

```python
sections, local_bears, global_bears, targets = gather_configuration(
            lambda *args: True,
            self.log_printer,
            arg_list=[])
```

notice how the first argument is a lambda function returning `True`.

this is `gather_configuration` as defined in `ConfigurationGathering.py`:

```python
def gather_configuration(acquire_settings,
                         log_printer,
                         autoapply=None,
                         arg_list=None,
                         arg_parser=None):
```

so `acquire_settings` is said lambda function returning `True` when executing this test.

`gather_configuration` doesn't alter or use `acquire_settings`, but calls

```python
local_bears, global_bears = fill_settings(sections,
                                          acquire_settings,
                                          log_printer)
```

which in turn doesn't alter or use `acquire_settings`, but calls:

```python
fill_section(section, acquire_settings, log_printer, all_bears)
```

In `fill_section`, defined in `SectionFilling.py`, things finally go downhill on it's first use of `acquire_setting` when it runs:

```python
if len(needed_settings) > 0:
    new_vals = acquire_settings(log_printer, needed_settings)
    for setting, help_text in new_vals.items():
        section.append(Setting(setting, help_text))
```

remember how `acquire_settings` is a lambda function returning `True`?
well... `new_vals` is `True` and `new_vals.items()` ... ouch.

```python
AttributeError: 'bool' object has no attribute 'items'
```

and now please tell me:

## Why on earth does this **not** fail on CI?

**Comment by fneu** [comment]
If we look i.e. at the [circleci run}(https://circleci.com/gh/coala-analyzer/coala/6631) on my [last pr](https://github.com/coala-analyzer/coala/pull/2209) from today:

one can see under *$ bash .misc/test.sh*:

```
tests/settings/ConfigurationGatheringTest.py ...........
```

which means that all ConfigurationGatheringTests get run and pass.
Not `F` for fail or `s` for skipped...

**Comment by Makman2** [comment]
duplicate of https://github.com/coala-analyzer/coala/issues/2111

closing the other one in favor of this because of your debug session?

**Comment by mtomwing** [comment]
TL;DR if we install coala-bears before running the tests then everything fails as expected.

It looks like `coala-bears` is only installed in the circle-ci plan AFTER the main coala tests are run. When I ran the tests without having `coala-bears` installed everything was grand.

I'm not too familiar with the coala codebase but by looking at:

```python
    # Strip away existent settings.
    needed_settings = {}
    for setting, help_text in prel_needed_settings.items():
        if not setting in section:
            needed_settings[setting] = help_text

    # Get missing ones.
    if len(needed_settings) > 0:
        new_vals = acquire_settings(log_printer, needed_settings)
        for setting, help_text in new_vals.items():
            section.append(Setting(setting, help_text))
```

It seems like if there are no bears found, then `needed_settings` is never populated with anything. This leads to the last block being skipped entirely.

**Comment by sils1297** [comment]
So we should make this test independent of the installed bears

**Comment by mtomwing** [comment]
@sils1297 agreed. Though at some point we should be testing that this branch does work.
